### PR TITLE
feat: add method for getting populated experiments

### PR DESCRIPTION
**Pull request by gbockus-sf**
### What does this PR do?
I wired up the experiments service with E4D just to verify things are working as expected.  Led me down a refactoring/extending path.  

Added method for: 
- getting populated experiments with status 
- getting single experiment status
- getting list of key/value (ExperimentState) values 

### What issues does this PR fix or reference?
@W-16328933@
@W-16328936@

### How to link with E4D
in this repo run 
1. npm install
2. npm compile
3. npm link
4. rm -rf node_modules
5. npm install --omit=dev

In the extension repo
1. npm link @salesforce/salesforcedx-vscode-experiments

Example usage
```


 const experiments: ExperimentDefinition[] = [{
    name: 'RAG-Test',
    type: ExperimentType.Stateful,
    distributionPercent: 50,
    expirationDate: '2024-12-31T23:59:59Z'
  },
{
  name: 'RAG-Test1',
  type: ExperimentType.Stateful,
  distributionPercent: 99,
  expirationDate: '2024-01-31T23:59:59Z'
},
{
  name: 'RAG-Test2',
  type: ExperimentType.Stateful,
  distributionPercent: 99,
  expirationDate: '2024-12-31T23:59:59Z'
}];

  const experimentService = getExperimentService();
  await experimentService.registerExperiments(context, experiments);
  const assignedExperiments = experimentService.getExperiments();
  const experimentsState = experimentService.getExperimentsState();
  const isOn = experimentService.getExperimentState(experiments[0]);
  console.log('***** Experiment ***** ', assignedExperiments);
  console.log('***** Experiment ***** ', isOn, experimentsState);
```

### CODE REVIEW

File: .github/workflows/automerge.yml [hunk 1]
```diff
@@ -7,4 +7,6 @@ on:
 jobs:
   automerge:
     uses: salesforcecli/github-workflows/.github/workflows/automerge.yml@main
+    with:
+      GITHUB_TOKEN: ${{ secrets.IDEE_GH_TOKEN }}
```
Comment (gbockus-sf) [inline .github/workflows/automerge.yml:11]: I think this will fix the dependabot automerge failures.

File: package.json [hunk 2]
```diff
@@ -43,6 +43,9 @@
     "typescript": "^5.4.2",
     "wireit": "^0.14.5"
   },
+  "peerDependencies": {
+    "@types/vscode": "^1.76.0"
+  },
```
Comment (gbockus-sf) [inline package.json:48]: When trying to test using E4D I was getting compile conflicts if vscode lib module was installed...realized this is a peer dep so added it.

File: src/api.ts [hunk 3]
```diff
@@ -29,6 +30,12 @@ export interface Experiment extends ExperimentDefinition {
 }
 
 export interface IExperimentService {
-  registerExperiments(context: vscode.ExtensionContext, experiments: ExperimentDefinition[]): void;
+  registerExperiments(context: ExtensionContext, experiments: ExperimentDefinition[]): Promise<void>;
   getExperiments(): Experiment[];
+  getExperimentsState(): ExperimentState
+  getExperimentState(experiment: ExperimentDefinition): boolean;
```
Comment (gbockus-sf) [inline src/api.ts:36]: Expanded to support use cases. 
Made registerExperiments return a promise since it has async calls. 
See comments in code for use cases.

File: src/api.ts [hunk 4]
```diff
@@ -29,6 +30,12 @@ export interface Experiment extends ExperimentDefinition {
...
 }
+
+export type ExperimentState = {
+  [key: string]: boolean;
+};
```
Comment (gbockus-sf) [inline src/api.ts:41]: returning this type so moved it here.

File: src/experimentService.ts [hunk 5]
```diff
@@ -6,11 +6,11 @@
...
 
 import * as vscode from 'vscode';
-import { ExperimentDefinition, Experiment, IExperimentService } from './api';
-import { getExperimentStatus } from './internals/utils';
+import { ExperimentDefinition, Experiment, IExperimentService, ExperimentState } from './api';
+import { ExperimentStateManager } from './internals/experimentState';
 
 class ExperimentService implements IExperimentService {
-  private experiments: ExperimentDefinition[] = [];
+  // private experiments: ExperimentDefinition[] = [];
```
Comment (gbockus-sf) [inline src/experimentService.ts:13]: I got some cleanup to do if this sticks.

File: src/internals/experimentState.ts [hunk 6]
```diff
@@ -28,21 +27,26 @@ export class ExperimentStateManager {
...
         if (isExpired(experiment.expirationDate)) {
-          this.stateCache[experiment.name] = false;
+          status = ExperimentStatus.Expired;
         }
+        this.stateCache[experiment.name] = status === ExperimentStatus.Active ? true : false;
+        return {...experiment, status};
       });
 
     await this.context.globalState.update(EXPERIMENT_STATE_KEY, this.stateCache);
+    this.experiments = populatedExperiments;
```
Comment (gbockus-sf) [inline src/internals/experimentState.ts:45]: I needed the ability to get back the list of updated Experiments so I added the additional instance var.

File: src/internals/experimentState.ts [hunk 7]
```diff
@@ -28,21 +27,26 @@ export class ExperimentStateManager {
...
   }
 
   getExperimentState(experiment: ExperimentDefinition): boolean {
+    // Should we adjust this on the fly or used the expired status found on load? 
     if (isExpired(experiment.expirationDate)) {
       return false;
     }
```
Comment (gbockus-sf) [inline src/internals/experimentState.ts:52]: Thinking about if we should not adjust the experiment on the fly, but use what's defined at startup.

File: tsconfig.json [hunk 8]
```diff
@@ -3,11 +3,12 @@
     "module": "Node16",
     "target": "ES2022",
     "lib": ["ES2023"],
-    "outDir": "out",
+    "outDir": "lib",
```
Comment (gbockus-sf) [inline tsconfig.json:6]: found some inconsistencies when using the lib.

File: .github/workflows/automerge.yml (hunk 1 above)
Comment (jgellin-sf) [reply .github/workflows/automerge.yml:11]: 👍

File: package.json (hunk 2 above)
Comment (jgellin-sf) [reply package.json:48]: 👍

File: src/api.ts (hunk 3 above)
Comment (jgellin-sf) [reply src/api.ts:36]: 👍

File: src/api.ts [hunk 9]
```diff
@@ -14,6 +14,7 @@ export enum ExperimentType {
 
 export enum ExperimentStatus {
   Active = 'active',
+  Disabled = 'disabled',
   Expired = 'expired'
 }
```
Comment (jgellin-sf) [inline src/api.ts:19]: My conception of `ExperimentStatus` was originally just to indicate whether the experiment was still a going concern ('active') or not ('expired'). It looks like you are using `disabled` as a synonym for a `false` state, and kind of the opposite of `active`. This is fine, but if we are going to do this, then:
1. We should define the state in 1 place and not 2, and get rid of the boolean state
2. We should change the names in the enum so that the current `active` and `disabled` statuses are true opposites (active/inactive or enabled/disabled).

File: src/api.ts (hunk 4 above)
Comment (jgellin-sf) [reply src/api.ts:41]: This is good to move it here, but see my other comments about the status/state conflict.

File: src/internals/experimentState.ts (hunk 7 above)
Comment (jgellin-sf) [reply src/internals/experimentState.ts:52]: Originally I added this if-expired clause for Transactional experiments, since they are not in the cache.

I thought about this on-the-fly question when you suggested I move this clause above the next one in the last PR. If the expired status was in the cache, then the order wouldn't matter, so at first I was reluctant to change the order, because it wouldn't matter. But then I thought maybe some customers leave their IDE open for weeks-long sessions like I do, and the experiment should expire when it's supposed to, so we _should_ check on the fly and then order does matter.

TLDR: Yes, on-the-fly

File: tsconfig.json (hunk 8 above)
Comment (jgellin-sf) [reply tsconfig.json:6]: 👍

**Review by jgellin-sf — COMMENTED** [review]
We should decide on the meaning of Status. If state is included in status, let's get rid of state.

**Review by jgellin-sf — COMMENTED** [review]
Almost there. A couple cleanup things.

File: src/api.ts [hunk 10]
```diff
@@ -26,9 +26,16 @@ export interface ExperimentDefinition {
 
 export interface Experiment extends ExperimentDefinition {
   status: ExperimentStatus;
+  state: boolean;
```
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply src/api.ts:29]: still processing :)

File: src/internals/experimentState.ts [hunk 11]
```diff
@@ -28,28 +25,44 @@ export class ExperimentStateManager {
...
+    // assign transactional experiments
     return randomAssignment(experiment);
   }
+
+  getExperimentsState(): typeof this.stateCache {
+    return this.stateCache;
+  }
+
+  getExperiments(): typeof this.experiments {
```
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply src/internals/experimentState.ts:65]: Doesn't bother me, but also not opposed to updating.  👍

File: src/internals/experimentState.ts [hunk 12]
```diff
@@ -28,28 +25,44 @@ export class ExperimentStateManager {
...
-    // check if the experiment has been assigned
+    // check if the experiment has been assigned, should always be true for Stateful experiments
     if (this.stateCache[experiment.name] !== undefined) {
       return this.stateCache[experiment.name];
     }
+    // assign transactional experiments
     return randomAssignment(experiment);
   }
+
+  getExperimentsState(): typeof this.stateCache {
```
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply src/internals/experimentState.ts:61]: Doesn't bother me, but also not opposed to updating. 👍

**Review by jgellin-sf — APPROVED** [review]
🚢

File: src/api.ts (hunk 9 above)
Comment (gbockus-sf) [reply src/api.ts]: we got to a _current_ resolution
